Patch series to reapply #118734 and substantially improve it

chandlerc opened the pull request:
This PR is for a patch series of 6 patches to try and fully address the Clang builtin string tables. I'm posting it as a single PR to give context to the series and explain what the overarching goal is, but each of the 6 patches is designed to be reasonable to review on its own and land on its own. However, intermediate states may push MSVC's compile time and memory usage due to especially large string tables.

The overarching goal of this patch series is to re-land the port of Clang's builtin tables to use offsets into string tables rather than containing dynamically relocated pointers to individual strings. A naive port as in #118734 has some serious issues due to forming *extraordinarily* large string tables, and many of them. To address this, subsequent patches restructure the builtin tables to make them smaller through sharding, use of TableGen-ed string tables without duplicates, and extracting a common prefix that can be dynamically added. The first patch results in a string table over 500KiB, but at the end the largest is well below 200KiB.

Combined, this patch series reduces the readonly data used by Clang's builtin tables by roughly 50% in addition to removing the dynamic relocations needed. The net result is a 30% reduction in the binary's dynamic relocation metadata, and a 2% reduction in the total binary size.

```
    FILE SIZE        VM SIZE
 --------------  --------------
  +0.8%  +405Ki  +0.8%  +405Ki    .rodata
  +0.0% +6.61Ki  +0.0% +6.61Ki    .text
  +0.0%    +300  +0.0%    +300    .dynstr
  +0.0%    +252  +0.0%    +252    .eh_frame
  +0.0%    +168  +0.0%    +168    .dynsym
  +0.0%     +48  +0.0%     +48    .eh_frame_hdr
  +0.0%     +36  +0.0%     +36    .gnu.hash
  +2.3%     +18  +2.3%     +18    [LOAD #2 [R]]
  +0.0%     +14  +0.0%     +14    .gnu.version
 -33.3%     -12  [ = ]       0    [Unmapped]
  [ = ]       0 -38.5%    -320    .relro_padding
 -28.9% -1.36Mi -28.9% -1.36Mi    .data.rel.ro
 -29.6% -2.57Mi -29.6% -2.57Mi    .rela.dyn
  -2.2% -3.53Mi  -2.2% -3.53Mi    TOTAL
```

An overview of the various patches in the series:

1) Reapplies #118734 but using the new ADT for representing the string tables and offsets It also restructures the builtin tables to support sharding both the main and target builtins into multiple shards. This both allows each shard to be smaller, and also allows independent construction of each shard. And last but not least, it supports factoring a common prefix out of a shard.
2) Leverages the new structure to switch RISCV (a target with larger tables) to directly generate some of its builtins data with TableGen rather than using X-macros. This lets TableGen generate the string table, eliminating redundant strings and reducing macro-based compile time memory overhead. It also factors out common prefixes.
3) Repeats the restructuring of (2) for AArch64 and ARM, which includes the largest builtin table (SVE).
4) Teach common builtin tablegen emission to generate string tables, updating all the targets that leverage this.
5) Add prefix factoring to common builtin tablegen and use it in the x86 target.
6) Use prefix factoring in the Hexagon builtins.

Each patch has a more detailed description in its commit message.

chandlerc committed:
[StrTable] Switch Clang builtins to use string tables

This both reapplies #118734, the initial attempt at this, and updates it
significantly.

First, it uses the newly added `StringTable` abstraction for string
tables, and simplifies the construction to build the string table and
info arrays separately. This should reduce any `constexpr` compile time
memory or CPU cost of the original PR while significantly improving the
APIs throughout.

It also restructures the builtins to support sharding across several
independent tables. This accomplishes two improvements from the
original PR:

1) It improves the APIs used significantly.

2) When builtins are defined from different sources (like SVE vs MVE in
   AArch64), this allows each of them to build their own string table
   independently rather than having to merge the string tables and info
   structures.

3) It allows each shard to factor out a common prefix, often cutting the
   size of the strings needed for the builtins by a factor two.

The second point is important both to allow different mechanisms of
construction (for example a `.def` file and a tablegen'ed `.inc` file,
or different tablegen'ed `.inc files), it also simply reduces the sizes
of these tables which is valuable given how large they are in some
cases. The third builds on that size reduction.

Initially, we use this new sharding rather than merging tables in
AArch64, LoongArch, RISCV, and X86. Mostly this helps ensure the system
works, as without further changes these still push scaling limits.
Subsequent commits will more deeply leverage the new structure,
including using the prefix capabilities which cannot be easily factored
out here and requires deep changes to the targets.

chandlerc committed:
[StrTable] Switch RISCV to leverage sharded, prefixed builtins w/ TableGen

This lets the TableGen-ed code be much cleaner, directly building an
efficient string table without duplicates and without the repeated
prefix.

chandlerc committed:
[StrTable] Switch AArch64 and ARM to use directly TableGen-ed builtin tables

This leverages the sharded structure of the builtins to make it easy to
directly tablegen most of the AArch64 and ARM builtins while still using
X-macros for a few edge cases. It also extracts common prefixes as part
of that.

This makes the string tables for these targets dramatically smaller.
This is especially important as the SVE builtins represent (by far) the
largest string table and largest builtin table across all the targets in
Clang.

chandlerc committed:
[StrTable] Teach main builtin TableGen to use direct enums, strings, and info

This moves the main builtins and several targets to use nice generated
string tables and info structures rather than X-macros. Even without
obvious prefixes factored out, the resulting tables are significantly
smaller and much cheaper to compile with out all the X-macro overhead.

This leaves the X-macros in place for atomic builtins which have a wide
range of uses that don't seem reasonable to fold into TableGen.

As future work, these should move to their own file (whether as X-macros
or just generated patterns) so the AST headers don't have to include all
the data for other builtins.

chandlerc commented:
@dyung - I believe this PR may be a credible path to address the issues hit with your MSVC builders, would appreciate any help testing it in advance if possible.

dyung commented:
> > @dyung - I believe this PR may be a credible path to address the issues hit with your MSVC builders, would appreciate any help testing it in advance if possible.
> 
> Sure, I'll give it a try

You seem to still be working on it, can you tell me which commit I should try building/testing when you are done?

chandlerc commented:
> > > @dyung - I believe this PR may be a credible path to address the issues hit with your MSVC builders, would appreciate any help testing it in advance if possible.
> > 
> > 
> > Sure, I'll give it a try
> 
> You seem to still be working on it, can you tell me which commit I should try building/testing when you are done?

Sorry, was just tidying up things that I missed until the PR was uploaded. Only rebasing on top-of-tree and a clang-format issue, nothing that changes functionality. And I think all done now so you should be good-to-go with the top of this PR (I think [`2bcc4e5` (#120534)](https://github.com/llvm/llvm-project/pull/120534/commits/2bcc4e5f7043dab1ef673dd20b38009363db51db) is the link to that commit)

dyung commented:
> > > > @dyung - I believe this PR may be a credible path to address the issues hit with your MSVC builders, would appreciate any help testing it in advance if possible.
> > > 
> > > 
> > > Sure, I'll give it a try
> > 
> > 
> > You seem to still be working on it, can you tell me which commit I should try building/testing when you are done?
> 
> Sorry, was just tidying up things that I missed until the PR was uploaded. Only rebasing on top-of-tree and a clang-format issue, nothing that changes functionality. And I think all done now so you should be good-to-go with the top of this PR (I think [`2bcc4e5` (#120534)](https://github.com/llvm/llvm-project/pull/120534/commits/2bcc4e5f7043dab1ef673dd20b38009363db51db) is the link to that commit)

I'll try it and let you know. Give me about an hour or so.

chandlerc commented:
> I'll try it and let you know. Give me about an hour or so.

Awesome! But no huge rush, mostly just hoping this happens to dodge whatever has been tripping up things here.

dyung commented:
> > I'll try it and let you know. Give me about an hour or so.
> 
> Awesome! But no huge rush, mostly just hoping this happens to dodge whatever has been tripping up things here.

Sorry for the delay, but the failures still seem to be present. :( (The tests are still running, but the amount of output being generated is slowing things down).

Just a heads up, we are probably going to be upgrading our internal machines and upstream bot to a later build of VS2019 to avoid this issue. I'm validating a newer build today and if all goes well, I am going to try and get it deployed on Friday.

erichkeane commented on clang/include/clang/Basic/Builtins.h:
```diff
@@ -68,23 +70,144 @@ enum ID {
+  // Translate lengths to offsets. We start past the initial empty string at
+  // offset zero.
+  unsigned Offset = 1;
+  for (auto &I : Infos) {
```
Think this is a case where we're not supposed to use 'auto'.

erichkeane commented on clang/include/clang/Basic/Builtins.h:
```diff
@@ -68,23 +70,144 @@ enum ID {
+#ifdef __clang__
+#define CLANG_BUILTIN_DETAIL_STR_TABLE(S)                                      \
+  _Pragma("clang diagnostic push")                                             \
+      _Pragma("clang diagnostic ignored \"-Woverlength-strings\"")             \
```
Did Clang-format really do this?  The tabbing in each level seems weird.

erichkeane commented on clang/lib/Basic/Targets/XCore.h:
```diff
@@ -43,7 +43,8 @@ class LLVM_LIBRARY_VISIBILITY XCoreTargetInfo : public TargetInfo {
                         MacroBuilder &Builder) const override;
 
-  ArrayRef<Builtin::Info> getTargetBuiltins() const override;
+  std::pair<const llvm::StringTable *, ArrayRef<Builtin::Info>>
```
This std::pair is used a million times in this patch, IMO this should either:
a- be its own type to give each name a little more context.
OR
b- at minimum, a type alias.

chandlerc replied:
Yeah. I mostly didn't invest in tidying this because I knew it was going to change. ;]

Are you OK with just letting the series clean this up, or would you prefer a type alias in the initial patch?

erichkeane replied:
The series fixing it up is fine.  Particularly if it stays in 1 PR, where this will end up getting 'squashed' anyway.

erichkeane reviewed (commented):
Just a review of the 1st commit, I apologize for the way in advance I'm going about this :) 

As a stand-in for Aaron this week I feel responsible for reviewing this, but 1st patch seems reasonable.  I cannot claim to fully understand what is going on, but from what I can see this seems like a net improvement.

SO far, just a handful of nits.

erichkeane commented on clang/include/clang/Basic/Builtins.h:
```diff
@@ -112,6 +112,16 @@ static constexpr std::array<Info, N> MakeInfos(std::array<Info, N> Infos) {
+/// single table across all builtins.
+struct InfosShard {
+  const llvm::StringTable *Strings;
+  llvm::ArrayRef<Info> Infos;
```
Huh... looks like this happened? I presume/hope this gets used everywhere we used a std::pair in the previous patch.

erichkeane reviewed (commented):
No new comments, and it looks like the SmallVector of InfosShards fixes one of the previous concerns I had.

erichkeane commented:
Done with my review (I skipped the last 3, which are to just convert individual targets, though I glanced to make sure I was ok with the approach).

Overall, I'm positive on this, and think this is beneficial.  If this is something we can get to settle (I recognize this is probably what you were talking about with the RFC to increase the 'required MSVC version'), I'm all for it.

ONly a handful of nits, 1 of which on the 1st patch I think was solved in the 2nd.

chandlerc commented:
> > > I'll try it and let you know. Give me about an hour or so.
> > 
> > Awesome! But no huge rush, mostly just hoping this happens to dodge whatever has been tripping up things here.
> 
> Sorry for the delay, but the failures still seem to be present. :( (The tests are still running, but the amount of output being generated is slowing things down).

Strange... Is it the *same* output??

Can you upload the output somewhere? 

> 
> Just a heads up, we are probably going to be upgrading our internal machines and upstream bot to a later build of VS2019 to avoid this issue. I'm validating a newer build today and if all goes well, I am going to try and get it deployed on Friday.

While this generally seems good, I'd like to avoid requiring even more MSVC upgrades, so interested in knowing how this is failing.

chandlerc commented:
> > > > I'll try it and let you know. Give me about an hour or so.
> > > 
> > > 
> > > Awesome! But no huge rush, mostly just hoping this happens to dodge whatever has been tripping up things here.
> > 
> > 
> > Sorry for the delay, but the failures still seem to be present. :( (The tests are still running, but the amount of output being generated is slowing things down).
> 
> Strange... Is it the _same_ output??
> 
> Can you upload the output somewhere?

Just to add some more details now that I've slept a bit...

Previously there were errors in AArch64 and RISCV -- it'll be really useful to know if those are the only errors with this patch, are there new ones, and especially if the RISCV errors go away that might be helpful. If the errors are *only* in AArch64, then I have ideas to further improve this.

Also, last time that @zmodem helped look at the `.obj` file for this, I think we ended up looking at the wrong one -- I don't think it is the `clang/lib/Basic/Builtins.obj` that would show signs of the issue, but `clang/lib/Basic/Targets/AArch64.obj` (and potentially the `RISCV.obj`). Either those `.obj` files should be clearly wrong (IE, some of the arrays truncated / zeroed /etc), or this is a mis-*link* not a mis-*compile*...

But my goal is to somewhat experimentally find the threshold below which things start working. With thin patch series there is an architecture to mechanically cut things down more and more if needed, but need some idea of the target I should aim for.

dyung commented:
> > > > > I'll try it and let you know. Give me about an hour or so.
> > > > 
> > > > 
> > > > Awesome! But no huge rush, mostly just hoping this happens to dodge whatever has been tripping up things here.
> > > 
> > > 
> > > Sorry for the delay, but the failures still seem to be present. :( (The tests are still running, but the amount of output being generated is slowing things down).
> > 
> > 
> > Strange... Is it the _same_ output??
> > Can you upload the output somewhere?
> 
> Just to add some more details now that I've slept a bit...
> 
> Previously there were errors in AArch64 and RISCV -- it'll be really useful to know if those are the only errors with this patch, are there new ones, and especially if the RISCV errors go away that might be helpful. If the errors are _only_ in AArch64, then I have ideas to further improve this.
> 
> Also, last time that @zmodem helped look at the `.obj` file for this, I think we ended up looking at the wrong one -- I don't think it is the `clang/lib/Basic/Builtins.obj` that would show signs of the issue, but `clang/lib/Basic/Targets/AArch64.obj` (and potentially the `RISCV.obj`). Either those `.obj` files should be clearly wrong (IE, some of the arrays truncated / zeroed /etc), or this is a mis-_link_ not a mis-_compile_...
> 
> But my goal is to somewhat experimentally find the threshold below which things start working. With this patch series there is an architecture to mechanically cut things down more and more if needed, but need some idea of the target I should aim for.

Sure. I deleted the object and log files, but can easily run it again. IIRC the tests that I saw failing in the console were LoongArch, but I think many/most targets were affected.

I'll take the buildbot offline, run it again and then archive the logs and object files somewhere.

dyung commented:
> > > > > > I'll try it and let you know. Give me about an hour or so.
> > > > > 
> > > > > 
> > > > > Awesome! But no huge rush, mostly just hoping this happens to dodge whatever has been tripping up things here.
> > > > 
> > > > 
> > > > Sorry for the delay, but the failures still seem to be present. :( (The tests are still running, but the amount of output being generated is slowing things down).
> > > 
> > > 
> > > Strange... Is it the _same_ output??
> > > Can you upload the output somewhere?
> > 
> > 
> > Just to add some more details now that I've slept a bit...
> > Previously there were errors in AArch64 and RISCV -- it'll be really useful to know if those are the only errors with this patch, are there new ones, and especially if the RISCV errors go away that might be helpful. If the errors are _only_ in AArch64, then I have ideas to further improve this.
> > Also, last time that @zmodem helped look at the `.obj` file for this, I think we ended up looking at the wrong one -- I don't think it is the `clang/lib/Basic/Builtins.obj` that would show signs of the issue, but `clang/lib/Basic/Targets/AArch64.obj` (and potentially the `RISCV.obj`). Either those `.obj` files should be clearly wrong (IE, some of the arrays truncated / zeroed /etc), or this is a mis-_link_ not a mis-_compile_...
> > But my goal is to somewhat experimentally find the threshold below which things start working. With this patch series there is an architecture to mechanically cut things down more and more if needed, but need some idea of the target I should aim for.
> 
> Sure. I deleted the object and log files, but can easily run it again. IIRC the tests that I saw failing in the console were LoongArch, but I think many/most targets were affected.
> 
> I'll take the buildbot offline, run it again and then archive the logs and object files somewhere.

Here is the logfile from the run. This is doing the exact same run as the buildbot (I ran the same commands as the buildbot would run). The compressed logfile is around 15MB, the uncompressed is around 300MB.

https://www.dropbox.com/scl/fi/urcanfwpf6da1dz2i5t85/log_test.zip?rlkey=5ludb4vpoe32mfp7hlnelrr47&dl=0

It's going to take a while to extract the build directory as it is 8GB in size, but if you have specific files you want to see let me know and I can pick them out.

Note that our bot doesn't build ARM/AArch64, so it would make sense that we don't see failures there (although I can't remember if we did previously...)

chandlerc commented:
> > > Just to add some more details now that I've slept a bit...
> > > Previously there were errors in AArch64 and RISCV -- it'll be really useful to know if those are the only errors with this patch, are there new ones, and especially if the RISCV errors go away that might be helpful. If the errors are _only_ in AArch64, then I have ideas to further improve this.
> > > Also, last time that @zmodem helped look at the `.obj` file for this, I think we ended up looking at the wrong one -- I don't think it is the `clang/lib/Basic/Builtins.obj` that would show signs of the issue, but `clang/lib/Basic/Targets/AArch64.obj` (and potentially the `RISCV.obj`). Either those `.obj` files should be clearly wrong (IE, some of the arrays truncated / zeroed /etc), or this is a mis-_link_ not a mis-_compile_...
> > > But my goal is to somewhat experimentally find the threshold below which things start working. With this patch series there is an architecture to mechanically cut things down more and more if needed, but need some idea of the target I should aim for.
> > 
> > 
> > Sure. I deleted the object and log files, but can easily run it again. IIRC the tests that I saw failing in the console were LoongArch, but I think many/most targets were affected.
> > I'll take the buildbot offline, run it again and then archive the logs and object files somewhere.
> 
> Here is the logfile from the run. This is doing the exact same run as the buildbot (I ran the same commands as the buildbot would run). The compressed logfile is around 15MB, the uncompressed is around 300MB.
> 
> [dropbox.com/scl/fi/urcanfwpf6da1dz2i5t85/log_test.zip?rlkey=5ludb4vpoe32mfp7hlnelrr47&dl=0](https://www.dropbox.com/scl/fi/urcanfwpf6da1dz2i5t85/log_test.zip?rlkey=5ludb4vpoe32mfp7hlnelrr47&dl=0)
> 
> It's going to take a while to extract the build directory as it is 8GB in size, but if you have specific files you want to see let me know and I can pick them out.
> 
> Note that our bot doesn't build ARM/AArch64, so it would make sense that we don't see failures there (although I can't remember if we did previously...)

Thanks, this is already really helpful.

The bots do seem to build the Clang parts of AArch64:

```
PASS: Clang :: AST/builtins-arm-strex-rettype.c (438 of 82164)
```

This was one the failures previously, and is now passing. I'm not seeing any of the AArch64, ARM, or RISCV failures that I was seeing. But there are still x86 and LoongArch failures. That may point me at how to make progress.

Let me work on adding a few more patches to the series.

chandlerc commented:
@dyung -- Could you try out https://github.com/chandlerc/llvm-project/tree/shard-loongarch and see if that works?

Notably, it includes one additional patch on top of this series: https://github.com/chandlerc/llvm-project/commit/2d593288dc18c55307779ae82a18d024761356ad

This won't be enough to fully resolve the issue, but if the LoongArch errors go away with this patch, then it gives me a strong theory about what is still going wrong and I can likely find a workaround for the rest as well.

dyung commented:
> @dyung -- Could you try out https://github.com/chandlerc/llvm-project/tree/shard-loongarch and see if that works?
> 
> Notably, it includes one additional patch on top of this series: [chandlerc@2d59328](https://github.com/chandlerc/llvm-project/commit/2d593288dc18c55307779ae82a18d024761356ad)
> 
> This won't be enough to fully resolve the issue, but if the LoongArch errors go away with this patch, then it gives me a strong theory about what is still going wrong and I can likely find a workaround for the rest as well.

Sorry for the delay, here are the results of running the branch shard-loongarch. Main takeaway is that the loongarch failures seem to have been fixed with the patch at the HEAD of that branch.

I have uploaded the test logfile here (another 300MB file when expanded):
https://www.dropbox.com/scl/fi/8v8toveo8h45c4iuwjgo8/log_test.f6426abc224b140bfb3b6fdc45443bc4e16057aa.zip?rlkey=6rj3vygil705ddd63vk4ykqdl&dl=0

chandlerc commented:
No worries about delay, this gives me a credible target to resolve the rest of the issues. I'll update this PR both to address review comments but also to try and address the rest of the failures. Appreciate runs to validate these updates. =]

dyung commented:
> No worries about delay, this gives me a credible target to resolve the rest of the issues. I'll update this PR both to address review comments but also to try and address the rest of the failures. Appreciate runs to validate these updates. =]

Definitely happy to help out with runs as we do not want our stuff broken internally! Just ping me and I'd be happy to run your changes on our build bot.

h-vetinari commented:
I've [tested](https://github.com/conda-forge/clangdev-feedstock/pull/333) this PR (as of 2bcc4e5f7043dab1ef673dd20b38009363db51db) in our infrastructure and can confirm that things run fine with VS2019 again. Thanks a lot for reworking this! 🙏

chandlerc reviewed (commented):
I think I've addressed most of the review comments here at this point.

But maybe most excitingly, I think the latest version may dodge the issues that have cropped up with MSVC -- both LoongArch and X86 fixes have been incorporated that hopefully help. @dyung -- if you can try it out, let me know.

There are three PRs factored out of this and under independent review:
- #120831
- #120835 (response to review comment)
- #120861

The patches that are still in this series because I don't really know how to send them out independently without breaking MSVC have some changes from the initial PR going out:

- Now sharding LoongArch as well as other targets in the PR that introduces sharding
- An extra commit in the series specifically moving the main builtins and several targets that use the same tablegen emission logic to not use X-macros as much. This is a big new commit and should get specific review in this PR before it lands: [`fcb7395` (#120534)](https://github.com/llvm/llvm-project/pull/120534/commits/fcb739514d298a2cf0b33d4dacfa2a2ce7277c68)

Beyond this, see responses to code reviwe comments inline.

dyung commented:
> I think I've addressed most of the review comments here at this point.
> 
> But maybe most excitingly, I think the latest version may dodge the issues that have cropped up with MSVC -- both LoongArch and X86 fixes have been incorporated that hopefully help. @dyung -- if you can try it out, let me know.

Sure, I can test it. Just to confirm, what branch/commit should I be testing?

chandlerc commented:
> > I think I've addressed most of the review comments here at this point.
> > But maybe most excitingly, I think the latest version may dodge the issues that have cropped up with MSVC -- both LoongArch and X86 fixes have been incorporated that hopefully help. @dyung -- if you can try it out, let me know.
> 
> Sure, I can test it. Just to confirm, what branch/commit should I be testing?

This PR has everything in it so you can just test it. There are 3 commits on this branch that won't land here (they're under review in their own PRs), but I've got them all here so you can just test this PR.

dyung commented:
> > Sure, I can test it. Just to confirm, what branch/commit should I be testing?
> 
> This PR has everything in it so you can just test it. There are 3 commits on this branch that won't land here (they're under review in their own PRs), but I've got them all here so you can just test this PR.

Unfortunely it still looks like there are a bunch of failures. They do look different from before though. I've uploaded the test log here: https://www.dropbox.com/scl/fi/d6e334heik4wp2dlztlq8/log_test.35a7fd79c471cdebd416ba711dbae6970e0e8550.zip?rlkey=kdkeo4897mw3642f9jyn4cmd0&dl=0

chandlerc commented:
> > > Sure, I can test it. Just to confirm, what branch/commit should I be testing?
> > 
> > 
> > This PR has everything in it so you can just test it. There are 3 commits on this branch that won't land here (they're under review in their own PRs), but I've got them all here so you can just test this PR.
> 
> Unfortunely it still looks like there are a bunch of failures. They do look different from before though. I've uploaded the test log here: [dropbox.com/scl/fi/d6e334heik4wp2dlztlq8/log_test.35a7fd79c471cdebd416ba711dbae6970e0e8550.zip?rlkey=kdkeo4897mw3642f9jyn4cmd0&dl=0](https://www.dropbox.com/scl/fi/d6e334heik4wp2dlztlq8/log_test.35a7fd79c471cdebd416ba711dbae6970e0e8550.zip?rlkey=kdkeo4897mw3642f9jyn4cmd0&dl=0)

Hmm, this looks like I just didn't fix "enough".

I've sent out #121043 and rebased this PR on top of that as well. Can you take another spin?

This PR should be the right branch, incorporating all the other changes.

dyung commented:
``
> Hmm, this looks like I just didn't fix "enough".
> 
> I've sent out #121043 and rebased this PR on top of that as well. Can you take another spin?
> 
> This PR should be the right branch, incorporating all the other changes.

I've built and tested this PR (I think) and it doesn't seem to have changed any from the last run?

Test logs are uploaded here: https://www.dropbox.com/scl/fi/icoi9z54vgayim9mueilg/log_test.979ae6b10db8a74ae3bff7f3f9fd34f535ce0910.zip?rlkey=up5j5kfgxttix4ajlakd1nee4&dl=0

chandlerc committed:
[StrTable] Add prefixes for x86 builtins.

This requires adding support to the general builtins emission for
producing prefixed builtin infos separately from un-prefixed which is
a bit crufty. But we don't currently have any good way of having a more
refined model than a single hard-coded prefix string per TableGen
emission. Something more powerful and/or elegant is possible, but this
is a fairly minimal first step that at least allows factoring out the
builtin prefix for something like X86.

chandlerc commented:
@dyung -- Ok, my latest attempt to work around these MSVC issues is now pushed to this PR. It also contains a commit of specifically debugging hacks to try and extract more information from any failure here. If you could try doing another build with the latest commit ([2ec750d](https://github.com/llvm/llvm-project/pull/120534/commits/2ec750d0105e1d4d52791f85293d16d28f6ccbd8)) in this PR, would be very interested in how it works.

dyung commented:
> @dyung -- Ok, my latest attempt to work around these MSVC issues is now pushed to this PR. It also contains a commit of specifically debugging hacks to try and extract more information from any failure here. If you could try doing another build with the latest commit ([2ec750d](https://github.com/llvm/llvm-project/pull/120534/commits/2ec750d0105e1d4d52791f85293d16d28f6ccbd8)) in this PR, would be very interested in how it works.

```
fatal error: error in backend: Name for __nvvm_vote_all is: ''!!!
```
I don't see any instances of the `__builtin_ia32_packsswb128` from a quick search of the test log.

Here is a compressed copy of the test log file:    [log_test.2ec750d0105e1d4d52791f85293d16d28f6ccbd8.zip](https://www.dropbox.com/scl/fi/60gizuhvi85lz8jxw44t9/log_test.2ec750d0105e1d4d52791f85293d16d28f6ccbd8.zip?rlkey=esic3xnyg4i11rjzuc9i0cr2a&dl=0)

chandlerc commented:
> > @dyung -- Ok, my latest attempt to work around these MSVC issues is now pushed to this PR. It also contains a commit of specifically debugging hacks to try and extract more information from any failure here. If you could try doing another build with the latest commit ([2ec750d](https://github.com/llvm/llvm-project/pull/120534/commits/2ec750d0105e1d4d52791f85293d16d28f6ccbd8)) in this PR, would be very interested in how it works.
> 
> ```
> fatal error: error in backend: Name for __nvvm_vote_all is: ''!!!
> ```
> I don't see any instances of the `__builtin_ia32_packsswb128` from a quick search of the test log.

This is awesome news!

It seems the last change was enough and gives me a strong guess as to the bug.

It does mean MSVC is miscompiling long string literals, so when you can upgrade, I would. And we should try to exclude these MSVC versions when we can.

But I may be able to work around more of these, stay tuned for some more updates.

> 
> Here is a compressed copy of the test log file:    [log_test.2ec750d0105e1d4d52791f85293d16d28f6ccbd8.zip](https://www.dropbox.com/scl/fi/60gizuhvi85lz8jxw44t9/log_test.2ec750d0105e1d4d52791f85293d16d28f6ccbd8.zip?rlkey=esic3xnyg4i11rjzuc9i0cr2a&dl=0)

chandlerc commented:
@dyung -- this PR is updated with new fixes. NVPTX hopefully works and neither of my debugging checks fires. But there may still be some other failures I need to chase down, let me know?

chandlerc commented:
> @dyung -- this PR is updated with new fixes. NVPTX hopefully works and neither of my debugging checks fires. But there may still be some other failures I need to chase down, let me know?

Hmm, looks like there are likely to be ARM and Hexagon failures remaining at least. An interesting question is whether there are any AArch64 failures as if so that will end all ideas I have to dodge this miscompile. I've now pushed a new update (cf4762f) that should help isolate exactly which targets continue to hit this.

dyung commented:
> > @dyung -- this PR is updated with new fixes. NVPTX hopefully works and neither of my debugging checks fires. But there may still be some other failures I need to chase down, let me know?
> 
> Hmm, looks like there are likely to be ARM and Hexagon failures remaining at least. An interesting question is whether there are any AArch64 failures as if so that will end all ideas I have to dodge this miscompile. I've now pushed a new update ([cf4762f](https://github.com/llvm/llvm-project/commit/cf4762f166d63fde810fd0e829e18e4a42e59e7b)) that should help isolate exactly which targets continue to hit this.

I've run your latest changes on the bot and I think there are still AArch64 failures as well as ARM failures as you predicted. I've put the test logfile up at https://www.dropbox.com/scl/fi/lh46zur1wlm4dz16ilooq/log_test.cf4762f166d63fde810fd0e829e18e4a42e59e7b.zip?rlkey=mcggerbkvq2rfkkisp34a3tds&dl=0

chandlerc commented:
> > > @dyung -- this PR is updated with new fixes. NVPTX hopefully works and neither of my debugging checks fires. But there may still be some other failures I need to chase down, let me know?
> > 
> > 
> > Hmm, looks like there are likely to be ARM and Hexagon failures remaining at least. An interesting question is whether there are any AArch64 failures as if so that will end all ideas I have to dodge this miscompile. I've now pushed a new update ([cf4762f](https://github.com/llvm/llvm-project/commit/cf4762f166d63fde810fd0e829e18e4a42e59e7b)) that should help isolate exactly which targets continue to hit this.
> 
> I've run your latest changes on the bot and I think there are still AArch64 failures as well as ARM failures as you predicted. I've put the test logfile up at [dropbox.com/scl/fi/lh46zur1wlm4dz16ilooq/log_test.cf4762f166d63fde810fd0e829e18e4a42e59e7b.zip?rlkey=mcggerbkvq2rfkkisp34a3tds&dl=0](https://www.dropbox.com/scl/fi/lh46zur1wlm4dz16ilooq/log_test.cf4762f166d63fde810fd0e829e18e4a42e59e7b.zip?rlkey=mcggerbkvq2rfkkisp34a3tds&dl=0)

:sigh: Yeah.

Well, that's pretty definitive. The SVE intrinsics names and types alone more than 2x the max size that this version of MSVC seems able to compile correctly. After the first 64k of string literal, it doesn't *error*, it just silently zeros the rest of the table. =[

Is there any hope of upgrading MSVC? I know you were looking at that, but not sure what progress happened there.

dyung commented:
> Is there any hope of upgrading MSVC? I know you were looking at that, but not sure what progress happened there.

I didn't go through with it and was hoping you would be able to find a work-around. I'll start talking to people to try and do a stop-gap update to a later version of VS2019 and let you know if/when we have deployed it. (I will be updating our upstream build bot as well at the same time).

chandlerc commented:
> > Is there any hope of upgrading MSVC? I know you were looking at that, but not sure what progress happened there.
> 
> I didn't go through with it and was hoping you would be able to find a work-around. I'll start talking to people to try and do a stop-gap update to a later version of VS2019 and let you know if/when we have deployed it. (I will be updating our upstream build bot as well at the same time).

What really baffles me is that LLVM itself should already be hitting this MSVC bug in serious ways starting with https://github.com/llvm/llvm-project/commit/381405fafe9d48d29c777e7680902d0943834859

That commit removed the only real workaround I know of from a *much* larger string table in the core of LLVM. What I can't figure out is why that string table compiles successfully but these don't. You all don't have any failing tests like `llvm/test/CodeGen/ARM/ldstrex.ll` or `llvm/test/Bitcode/arm-intrinsics.ll`?

Or any workaround patches to `llvm/lib/IR/Intrinsics.cpp`?

And line 18226-ish of the generated file `include/llvm/IR/IntrinsicImpl.inc`, you have `  98568, // llvm.arm.ldrex`? (Or if not that exact number, something larger than 64k...)

Mostly the above is for my curiosity to try and understand why the file in LLVM works despite all these in Clang not working....

I can basically revert parts of the above commit and make some other tablegen changes and I *should* be able to work around this if necessary. But it will make the above LLVM file and all these parts of Clang a bit slower to compile, sadly for all builds. So I'd like to avoid that if there is a reasonable way to upgrade. But if folks need to keep using this release series of MSVC, then we should somewhat systematically avoid risking this miscompile, even in the one place where it seems to work today.

chandlerc committed:
[StrTable] Add factored prefix for Hexagon

This target's builtins have an especially long prefix and so we get over
2x reduction in string table size required with this change.

chandlerc commented:
@dyung - OK, I think the current just-pushed version of this PR is worth another test.

I've taught the TableGen string table emission to go back to working around the MSVC issues using a different table form that we used to use in LLVM when MSVC had a reliable error on it. It appears that now it only sometimes gets miscompiled, and so this workaround got accidentally removed without really being in a reliable place.

But that still requires that all of the large string tables go through TableGen to get emitted in this form. So I've sent two PRs (separately, they seem independently good) to move NVPTX and Hexagon to TableGen, and this PR contains patches to move more of the ARM builtins to TableGen.

I *think* this gets all of the ones that produced too-large string tables, but testing should show. Let me know. If this works and there isn't a really short path to upgrade, I'll start pulling some of the pre-requisite PRs out for independent review, and then update this PR once I'm down to the minimal chain of PRs that do the full conversion of builtins to string tables.

Thanks,
-Chandler

dyung commented:
> @dyung - OK, I think the current just-pushed version of this PR is worth another test.
> 
> I've taught the TableGen string table emission to go back to working around the MSVC issues using a different table form that we used to use in LLVM when MSVC had a reliable error on it. It appears that now it only sometimes gets miscompiled, and so this workaround got accidentally removed without really being in a reliable place.
> 
> But that still requires that all of the large string tables go through TableGen to get emitted in this form. So I've sent two PRs (separately, they seem independently good) to move NVPTX and Hexagon to TableGen, and this PR contains patches to move more of the ARM builtins to TableGen.
> 
> I _think_ this gets all of the ones that produced too-large string tables, but testing should show. Let me know. If this works and there isn't a really short path to upgrade, I'll start pulling some of the pre-requisite PRs out for independent review, and then update this PR once I'm down to the minimal chain of PRs that do the full conversion of builtins to string tables.
> 
> Thanks, -Chandler

Some good news, everything seems to pass after your latest changes in this PR! I didn't believe it at first and did a clean rebuild and test to verify. In the end everything passed again.

That being said, I am working on deploying an updated version of VS2019 to our internal builders. I have gotten agreement with the plan, the only possible hold-up is that we would need to rebuild our internal builders which likely may not get done for a few weeks. I can apply quick work-arounds to the machines, but they ultimately will need to be rebuilt.

Would you prefer to move forward with the solution currently implemented in this PR or for us to update our internal builders?

chandlerc commented:
All of the dependent changes have landed, and I've re-structured this patch series to hopefully be easier to review. I still don't have a good way to land this incrementally, as the tools to work around MSVC limitations are only available when using TableGen-ed string tables, and those aren't available at first.

I could split out some of the later patches, but those are smaller and probably doesn't help as much. But the patches should be meaningful chunks to review here. There have been quite a few changes, so this does probably warrant a re-review from folks. Sorry for that.

@dyung -- as always, appreciate help verifying that this isn't going to break MSVC builds for you. The top of this PR should be up to date and ready to test.

dyung commented:
> All of the dependent changes have landed, and I've re-structured this patch series to hopefully be easier to review. I still don't have a good way to land this incrementally, as the tools to work around MSVC limitations are only available when using TableGen-ed string tables, and those aren't available at first.
> 
> I could split out some of the later patches, but those are smaller and probably doesn't help as much. But the patches should be meaningful chunks to review here. There have been quite a few changes, so this does probably warrant a re-review from folks. Sorry for that.
> 
> @dyung -- as always, appreciate help verifying that this isn't going to break MSVC builds for you. The top of this PR should be up to date and ready to test.

Happy to do so! Some updates on our end, I *think* that we have finished updating our internal machines with a newer build of Visual Studio 2019, but I need to confirm it. Our upstream bot has not yet been updated though, but I can do that pretty quickly/easily. I will test your changes on both versions tomorrow and let you know the results.

dyung commented:
> > All of the dependent changes have landed, and I've re-structured this patch series to hopefully be easier to review. I still don't have a good way to land this incrementally, as the tools to work around MSVC limitations are only available when using TableGen-ed string tables, and those aren't available at first.
> > I could split out some of the later patches, but those are smaller and probably doesn't help as much. But the patches should be meaningful chunks to review here. There have been quite a few changes, so this does probably warrant a re-review from folks. Sorry for that.
> > @dyung -- as always, appreciate help verifying that this isn't going to break MSVC builds for you. The top of this PR should be up to date and ready to test.
> 
> Happy to do so! Some updates on our end, I _think_ that we have finished updating our internal machines with a newer build of Visual Studio 2019, but I need to confirm it. Our upstream bot has not yet been updated though, but I can do that pretty quickly/easily. I will test your changes on both versions tomorrow and let you know the results.

I ran the most recent changes in this PR (2b85083ff8b85c6ebe62abe994cd08df61a9268c) and can confirm that all tests passed on both the original and newer version of Visual Studio 2019 that we are upgrading to.

chandlerc commented:
I know it's only been a few days, but pinging in the hope of landing this week... This seems to finally be in a good state and is somewhat hard to keep rebasing. Happy to do anything I can to help make review easier.

h-vetinari commented:
You'd have to merge and request a backport pretty quickly. Historically features could still be argued to land between rc1 (which is imminent) and rc2, but the windows is definitely closing fast

chandlerc commented:
Given the benefit to binary size and compile time in things like `configure` scripts, I'd certainly like to see it land...

Erich already reviewed an earlier version. Maybe @rnk can help with reviews?

rnk commented on clang/utils/TableGen/ClangBuiltinsEmitter.cpp:
```diff
@@ -482,17 +488,42 @@ void clang::EmitClangBuiltins(const RecordKeeper &Records, raw_ostream &OS) {
 
+  // Now collect (and count) the prefixed builtins.
+  unsigned NumPrefixedBuiltins = Builtins.size();
+  const Record *first_prefix = nullptr;
```
Super nit, just to prove I read all the code: Why the change in style here? `FirstPrefix`?

chandlerc replied:
Doh, just an accident, sorry. Out of the habits of LLVM style, fixed.

rnk reviewed (approved):
Thanks for optimizing the builtins! I feel like builtins have grown significantly since adding RISC-V and ARM MVE intrinsics, and few people until now have stopped to re-evaluate how we represent these things.

While I was watching these clearly generated Intrinsic files land in Clang, I was thinking, these are clearly generated from a data model using a script. Wouldn't it be more efficient to encode the underlying data model for these operations using a more normalized representation? I proposed this at one point, and someone from Intel assured me that these long repetitive instructions and intrinsics are hand-maintained. It must be true at some level. Regardless, this change seems like a positive step towards renormalizing the otherwise denormalized builtins databse.

There's a lot of code and I'm sure there are more nits I could pick out, but I think it makes sense to reland this with the more sharded approach and get some feedback in tree.

chandlerc commented:
Thanks @rnk !

I've fixed the one style nit (doh!) and a few surrounding variables.

I'm working on rebasing everything now.

But one specific question: would you prefer me to land as a series of commits or a single squashed commit for the entire PR? I'm happy either way. My mild preference is to prefer the series of commits, but open to suggestions here.

rnk commented:
> But one specific question: would you prefer me to land as a series of commits or a single squashed commit for the entire PR? I'm happy either way. My mild preference is to prefer the series of commits, but open to suggestions here.

I would land it as the six-patch stack in this case. I ended up using that to review, since mainly the first patch is the interesting one and the target-specific migration ones are more mechanical and more incremental.

chandlerc commented:
Merged:
- https://github.com/llvm/llvm-project/commit/cd269fee05a0f78fb53b65f701b4e06e9ddab424
- https://github.com/llvm/llvm-project/commit/1cb979f001b24c661b7d7adf50d7c9cf8adc593a
- https://github.com/llvm/llvm-project/commit/64ea3f5a4720105d166b034d5a34d92475579e64
- https://github.com/llvm/llvm-project/commit/212ecb9d5caaa7cc721edd981f36384ddfccfa5d
- https://github.com/llvm/llvm-project/commit/2ff42bdac3b9a131ce1c652d08edded4eac9d3f7
- https://github.com/llvm/llvm-project/commit/51d0ad7de0ad4636ae39783469cf555a1392b4ea

h-vetinari commented:
If you want to backport this, you need to add the PR to the milestone and comment 

`/cherry-pick <hash1> <hash2> <hash3> ...`

Then the bot should open a PR (or it'll tell you if it failed to cherry-pick, and how to proceed from there).
